students/alex_skrn/lesson06/mailroom.py

```diff
@@ -0,0 +1,273 @@
+#!/usr/bin/env python3
+
+"""Mailroom - Part 4 - Unit testing."""
+import os
+import datetime
+import tkinter as tk
+from tkinter import filedialog
+
+donors = {'Aristarkh Lentulov': [4.5, 5.0],
+          'El Lissitzky': [34.2, 30.0, 35.5],
+          'Kazimir Malevich': [15.0, 20.25, 12.25],
+          'Marc Chagall': [148.75, 155.0],
+          'Wassily Kandinsky': [75.0, 50.5, 60.4],
+          }
+
+
+# DONOR-RELATED FUNCTIONS
+def add_donation(name, amount):
+    """Add a donation for an existing or newly created donor."""
+    try:
+        donors[name].append(amount)
+    except KeyError:
+        donors[name] = [amount]
+
+
+def get_last_donation(name):
+    """Return a float -- the last donation of the given donor."""
+    return donors[name][-1]
+
+
+def get_donations(name):
+    """Return a list of the specified donor's donations."""
+    return donors[name]
+
+
+def get_total_given(name):
+    """Return total amount of donations for the given donor."""
+    return sum(get_donations(name))
+
+
+def sort_donors_by_total():
+    """Return a list of donor names sorted by total donations, max to min."""
+    donors_L = list(donors.items())
+    donors_sorted = sorted(donors_L, key=lambda x: sum(x[1]), reverse=True)
+    sorted_donor_names = [item[0] for item in donors_sorted]
+
+    return sorted_donor_names
+
+
+def print_donor_names():
+    """Print existing donor names on screen in alphabetical order."""
+    donors_L = sorted(donors.keys())
+    num = len(donors_L)
+    donors_S = (("\n" + ", ".join(["{}"] * num)).format(*donors_L))
+    print(donors_S)
+
+
+def get_email(name, amount):
+    """Return a str containing a thank-you email."""
+    email_text = ("""\nDear {},\n
+                  \nI would like to thank you for your donation of ${}.\n
+                  \nWe appreciate your support.\n
+                  \nSincerely,\n
+                  \nThe Organization\n
+                  """)
+    return email_text.format(name, amount)
+
+
+def create_report():
+    """Create and print a report."""
+    report = ""
+    title_line_form = "\n{:<26}{:^3}{:>13}{:^3}{:>13}{:^3}{:>13}\n"
+    title_line_text = ('Donor Name', '|', 'Total Given', '|',
+                       'Num Gifts', '|', 'Average Gift'
+                       )
+    report += title_line_form.format(*title_line_text)
+    report += str('- ' * 38)
+    form_line = "\n{:<26}{:>3}{:>13}{:>3}{:>13}{:>3}{:>13}"
+    for name in sort_donors_by_total():
+        total = get_total_given(name)
+        num_gifts = len(get_donations(name))
+        mean = round((total / num_gifts), 2)
+        report += (form_line.format(str(name), '$', str(total), ' ',
```

Comment:
no need to cast these as strings.  also, you can probably just put their logic inside the `.format()` instead of creating variables for each one.  Less overhead.